content/posts/replacing-the-redisgraph-parser.md

```diff
@@ -0,0 +1,109 @@
+
+## The joy of FOSS
+libcypher-parser was a perfect fit for our needs. Beyond the core criteria of being a feature-rich Cypher parser, it is a C implementation, stable but evolving, and - above all - open-source.
+
+RedisGraph is a better project for the issues opened, features requested, and contributions made by our open-source users. While integrating libcypher-parser, we've contributed code and feedback upstream, and collaboration that improves both projects is ongoing.
+
+The lesson that a vibrant open-source community enriches us all is not a new one, but I always find it heartening to be reminded of it!
+
+As an aside, it is interesting to note that instead of building a LALR parser (as Lemon does), libcypher-parser uses a PEG grammar. I won't go into the implications of this here, but Guido van Rossum has been writing a [very interesting series](https://medium.com/@gvanrossum_83706/peg-parsers-7ed72462f97c) on the subject for the curious.
+
+## Paid back with interest
+From start to finish, this undertaking took about 3 months, though even now we have open PRs to better leverage the new system. We now work with an immutable (though annotatable) AST, and have radically modified our access patterns and AST translation layer.
+
+To torture the metaphor of technical debt a little more, we recouped more than the principal over the course of this refactor. libcypher-parser gave us automatic support for a  number of features, ranging from syntactic sugar like implicit edges and partially-specified traversal ranges:
+```console
+MATCH (node1)-->(node2), (src)-[*2..]->(two_or_more_hops)
+```
+To more robust query validations, built-in precedence management for operators and comparisons, and handling for more edge cases than I can recall.
+
+We also took advantage of this moment to migrate all parsing to RedisGraph threads, so multiple queries can be parsed in parallel and the Redis server never blocks to perform this work.
+
+## The strength of strong abstractions
+Beyond these improvements, the greatest benefit of this refactor has been a much stronger abstraction layer between the AST and the tree of operations required to act upon it.
+
+Our earlier grammar made it easy to rely directly on the AST for the construction of operations, especially in superficially simple clauses like CREATE. This approach belies the complexity of circumstances that need to be considered, however. For example, given the query:
+```console
+MATCH (p:Person {name: 'Jeffrey'})
+CREATE (p)-[:EMPLOYED_BY]->(c: Company {name: 'Redis Labs'})
+RETURN *
+```
+The CREATE operation does not exist in isolation; it relies on projections from previous operations and itself  projects data that must be returned to the user.
```

Comment:
Change to: "The CREATE operation does not exist in isolation—it relies on projections from previous operations and itself  projects data that must be returned to the user."